pandas/core/indexes/multi.py

```diff
@@ -2163,11 +2164,41 @@ def append(self, other):
             (isinstance(o, MultiIndex) and o.nlevels >= self.nlevels) for o in other
         ):
             arrays = []
-            for i in range(self.nlevels):
-                label = self._get_level_values(i)
-                appended = [o._get_level_values(i) for o in other]
-                arrays.append(label.append(appended))
-            return MultiIndex.from_arrays(arrays, names=self.names)
+            if self.names.count(None) > 1 or any(
+                o.names.count(None) > 1 for o in other
+            ):
+
+                for i in range(self.nlevels):
+
+                    label = self._get_level_values(i)
+                    appended = [o._get_level_values(i) for o in other]
+                    arrays.append(label.append(appended))
+                    index_label_list = self.names
+
+            else:
+                index_label_list = self.get_unique_indexes(other)
+
```

Comment:
In this new method, a list of unique column name indexes from self and other are firstly obtained.